src/urllib3/response.py

```diff
@@ -701,14 +702,46 @@ def read(
         flush_decoder = False
         fp_closed = getattr(self._fp, "closed", False)
 
+        # Reading more than 2 GiB (`int` max value in C) via SSL on some
+        # CPython 3.8, 3.9, and 3.10 versions leads to an overflow error
+        # that has to be prevented if `amt` or `self.length_remaining`
+        # indicate that it may happen.
+        c_int_max = (2**31) - 1
+        if (
+            (amt and amt > c_int_max)
+            or (self.length_remaining and self.length_remaining > c_int_max)
+        ) and (
+            (3, 8, 0, "a", 4) <= sys.version_info < (3, 9, 7)
+            or (3, 10, 0, "b", 1) <= sys.version_info < (3, 10, 0, "rc", 1)
+        ):
+
+            def read(amt: Optional[int] = None) -> bytes:
+                if self._fp is None:
+                    return b""
+                buffer = io.BytesIO()
```

Comment:
Since we only use `buffer.write()` and `buffer.getvalue()`, why not use a simple `bytearray`?

Reply:
It looks like `BytesIO` is consistently faster than an unbounded `bytearray`?

```
$ python -m timeit -s 'import io; x = io.BytesIO()' 'for _ in range(10000):
    x.write(b"x" * 10000)

x.getvalue()
'
5 loops, best of 5: 45.9 msec per loop

$ python -m timeit -s 'import io; x = bytearray()' 'for _ in range(10000):
    x += b"x" * 10000

bytes(x)
'
2 loops, best of 5: 121 msec per loop
```

If that's the case we should use `io.BytesIO`, I wouldn't have guessed that!